File: report-generator/src/main/java/com/teamscale/report/util/BashFileSkippingInputStream.java
package com.teamscale.report.util;

import java.io.BufferedInputStream;
import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * Handles executable spring-boot jar files that prepend a bash file to the beginning of the ZIP file to make it
 * directly executable without "java -jar my.jar". We just skip the bash file until we find the zip file header.
 */
public class BashFileSkippingInputStream extends FilterInputStream {

	/**
	 * Wraps the given input stream in a BufferedInputStream and consumes all bytes until a zip file header is found.
	 */
	public BashFileSkippingInputStream(InputStream in) throws IOException {
		super(new BufferedInputStream(in));
		consumeUntilZipHeader();
	}

	/**
	 * Reads the stream until the zip file header "50 4B 03 04" is found or EOF is reached.
	 * After calling the method the read pointer points to the first byte of the zip file header.
	 */
	private void consumeUntilZipHeader() throws IOException {
		byte[] buffer = new byte[8192];
		in.mark(buffer.length);
		int count = in.read(buffer, 0, buffer.length);
		while (count > 0) {
			for (int i = 0; i < count - 3; i++) {
				if (buffer[i] == 0x50 && buffer[i + 1] == 0x4B && buffer[i + 2] == 0x03 && buffer[i + 3] == 0x04) {
					in.reset();
					in.skip(i);
					return;
				}
			}

			// Reset mark to 3 bytes before the end of the previously read buffer end to
			// also detect a zip header when it spans over two buffers
			in.reset();
			in.skip(buffer.length - 3);
			in.mark(buffer.length);
			count = in.read(buffer, 0, buffer.length);
		}
	}
}

File: report-generator/src/main/kotlin/com/teamscale/report/util/BashFileSkippingInputStream.kt
package com.teamscale.report.util

import java.io.BufferedInputStream
import java.io.FilterInputStream
import java.io.IOException
import java.io.InputStream

/**
 * This InputStream skips any initial bash script prepended to an executable spring-boot jar file,
 * positioning the stream at the start of the ZIP file header. This allows direct execution of
 * the jar without requiring "java -jar my.jar".
 */
class BashFileSkippingInputStream(input: InputStream) : FilterInputStream(BufferedInputStream(input)) {

	companion object {
		private val ZIP_HEADER = byteArrayOf(0x50, 0x4B, 0x03, 0x04)
		private const val BUFFER_SIZE = 8192
	}

	init {
		skipToZipHeader()
	}

	/**
	 * Reads the stream until the ZIP file header (0x50 4B 03 04) is found, or EOF is reached.
	 * After calling this method, the read pointer is positioned at the first byte of the ZIP header.
	 */
	@Throws(IOException::class)
	private fun skipToZipHeader() {
		val buffer = ByteArray(BUFFER_SIZE)
		`in`.mark(BUFFER_SIZE)

		var bytesRead = `in`.read(buffer, 0, BUFFER_SIZE)
		while (bytesRead != -1) {
			val headerIndex = findZipHeader(buffer, bytesRead)
			if (headerIndex != -1) {
				// Reset and skip to the start of the ZIP header
				`in`.reset()
				`in`.skip(headerIndex.toLong())
				return
			}

			// Adjust position and re-mark to check the buffer boundary for header
			`in`.reset()
			`in`.skip((BUFFER_SIZE - ZIP_HEADER.size + 1).toLong())
			`in`.mark(BUFFER_SIZE)
			bytesRead = `in`.read(buffer, 0, BUFFER_SIZE)
		}

		throw IOException("ZIP header not found in the input stream.")
	}

	/**
	 * Searches the buffer for the ZIP header signature.
	 * @param buffer The buffer to search.
	 * @param length The number of valid bytes in the buffer.
	 * @return The index where the ZIP header starts, or -1 if not found.
	 */
	private fun findZipHeader(buffer: ByteArray, length: Int) =
		(0..length - ZIP_HEADER.size)
			.firstOrNull {
				buffer[it] == ZIP_HEADER[0]
						&& buffer[it + 1] == ZIP_HEADER[1]
						&& buffer[it + 2] == ZIP_HEADER[2]
						&& buffer[it + 3] == ZIP_HEADER[3]
			} ?: -1
}
